### PR TITLE
Better area counting, valid action mask, human play interface for suite_go

### PR DESCRIPTION
**Pull request by emailweixu**
1. Use Tromp-Taylor rule for calculating area https://en.wikibooks.org/wiki/Computer_Go/Tromp-Taylor_Rules
2. Added a function to calculate all valid actions. The learning algorithm can use it to mask out invalid action
3. Added reward shaping to use softsign(area0 - area1 - winning_thresh) as reward.
All of the above 3 points are supposed to make the learning easier

4. Also added an interface to allow human to play with the agent. It can be invoked by
```python
python -m alf.bin.play --root_dir ... --gin_param GoEnvironment.human_player=0|1
```

### CODE REVIEW

File: alf/environments/suite_go.py [hunk 1]
```diff
@@ -62,13 +64,24 @@ class GoBoard(object):
...
+                 width,
+                 max_num_moves,
+                 num_previous_boards=10):
         """
         Args:
             batch_size (int): the number of parallel boards
             height (int): height of each board
             width (int): width of each board
+            max_num_moves (int): maximum number of moves allowed
+            num_previous_boards (int):
```
Comment (runjerry) [inline alf/environments/suite_go.py]: Maybe add comments here to give some brief idea why introducing this argument?

File: alf/environments/suite_go.py [hunk 2]
```diff
@@ -232,12 +262,60 @@ def calc_area(self, board_indices=None):
...
 
+    def calc_area(self, board_indices=None):
+        """Calculate the area of each player.
+
+        In order for a position to be considered to be owned by a player, it has
+        to be either the player's stone or cannot be reached by the opponent's stones.
+        With this definition of area, players have to play until all the dead stones
+        have been taken out. This shouldn't change how the game is played. This
+        is the so called `Tromp-Taylor rules <https://en.wikibooks.org/wiki/Computer_Go/Tromp-Taylor_Rules>`_
+
```
Comment (runjerry) [inline alf/environments/suite_go.py:282]: It would be very helpful if some more details regarding the algorithm implemented in line 295-309 can be added into the comments

File: alf/environments/suite_go.py [hunk 3]
```diff
@@ -256,81 +334,229 @@ def reset_board(self, board_indices=None):
...
+                          reshape(1, -1, 1, 1),
+                          self._y.unsqueeze(0),
+                          self._x.unsqueeze(0)] = 0
 
+    def classify_all_moves(self, player, board_indices=None):
+        """Classify all the moves on the board.
 
+        This function will examine all possible moves except PASS and annotate
+        them using 3 boolean attributes: occupied, suicidal, and repeated.
+
```
Comment (runjerry) [inline alf/environments/suite_go.py:360]: Maybe briefly mention its similarity/relationship to the update function?

File: alf/environments/suite_go.py [hunk 4]
```diff
@@ -367,138 +597,257 @@ def _reset(self):
...
+        invalid = occupied | repeated
+        if not self._allow_suicical_move:
+            invalid = invalid | suicidal
+
+        valid = (~invalid).reshape(self._batch_size, -1)
+        valid = torch.cat(
+            [valid, torch.ones(self._batch_size, 1, dtype=torch.bool)], dim=1)
+        return valid
+
+    def _step1(self, action):
```
Comment (runjerry) [inline alf/environments/suite_go.py:651]: Some comments to help understanding its relation with _step().

File: alf/environments/suite_go_test.py [hunk 5]
```diff
@@ -275,21 +387,81 @@ def test_environment(self):
...
-        self.assertEqual(time_step.reward, torch.tensor([0., 0., 1.]))
+        # player 1 wins
+        self.assertEqual(time_step.reward, torch.tensor([0., 0., -1.]))
         # board does not change after pass
-        self.assertEqual(time_step.observation['board'][2],
-                         torch.tensor(sequences[2][-1][2]).unsqueeze(0))
+        self.assertEqual(time_step.observation['board'][2][0],
+                         sequences[2][-1]['board'])
+
+    def test_environment2(self):
```
Comment (runjerry) [inline alf/environments/suite_go_test.py:406]: Maybe some comments here (and also test_environment) to tell the difference between to environment tests.

File: alf/environments/suite_go_test.py [hunk 6]
```diff
@@ -17,217 +17,321 @@
 
 import alf
 from alf.data_structures import StepType
-from alf.environments.suite_go import GoBoard, load
+from alf.environments.suite_go import GoBoard, GoEnvironment
 
 
 class BoardTest(alf.test.TestCase):
     def _get_test_cases(self):
```
Comment (runjerry) [inline alf/environments/suite_go_test.py:24]: Maybe briefly comment the different purposes of the three test sequences.

File: alf/environments/suite_go.py (hunk 1 above)
Comment (emailweixu) [reply alf/environments/suite_go.py]: added

File: alf/environments/suite_go.py (hunk 2 above)
Comment (emailweixu) [reply alf/environments/suite_go.py:282]: comments added

File: alf/environments/suite_go.py (hunk 3 above)
Comment (emailweixu) [reply alf/environments/suite_go.py:360]: added.

File: alf/environments/suite_go.py (hunk 4 above)
Comment (emailweixu) [reply alf/environments/suite_go.py:651]: added.

File: alf/environments/suite_go_test.py (hunk 6 above)
Comment (emailweixu) [reply alf/environments/suite_go_test.py:24]: Not much difference in terms of purpose. Just get more variation to make the test more robust.

File: alf/environments/suite_go_test.py (hunk 5 above)
Comment (emailweixu) [reply alf/environments/suite_go_test.py:406]: added.

File: alf/environments/suite_go.py [hunk 7]
```diff
@@ -367,138 +614,266 @@ def _reset(self):
...
+        """``_step()`` is used for actually update the board and calculate the
```
Comment (runjerry) [inline alf/environments/suite_go.py]: Do you mean _step1()?

**Review by runjerry — APPROVED** [review]
